provider/cluster/types/types.go

```diff
@@ -71,3 +75,70 @@ type ServiceLog struct {
 	Stream  io.ReadCloser
 	Scanner *bufio.Scanner
 }
+
+type LeaseEventObject struct {
+	Kind      string `json:"kind" yaml:"kind"`
+	Namespace string `json:"namespace" yaml:"namespace"`
+	Name      string `json:"name" yaml:"name"`
+}
+
+type LeaseEvent struct {
+	Type                string           `json:"type" yaml:"type"`
+	ReportingController string           `json:"reportingController,omitempty" yaml:"reportingController"`
+	ReportingInstance   string           `json:"reportingInstance,omitempty" yaml:"reportingInstance"`
+	Time                time.Time        `json:"time" yaml:"time"`
+	Reason              string           `json:"reason" yaml:"reason"`
+	Object              LeaseEventObject `json:"object" yaml:"object"`
+}
+
+// EventsWatcher
+type EventsWatcher interface {
+	Stop()
+	ResultChan() <-chan *eventsv1.Event
+}
+
+// EventsFeed
+type EventsFeed struct {
+	ctx    context.Context
+	cancel func()
+	quit   chan struct{}
+	feed   chan *eventsv1.Event
+}
+
+var _ EventsWatcher = (*EventsFeed)(nil)
+
+func NewEventsFeed(ctx context.Context) *EventsFeed {
+	ctx, cancel := context.WithCancel(ctx)
+	return &EventsFeed{
+		ctx:    ctx,
+		cancel: cancel,
+		quit:   make(chan struct{}),
+		feed:   make(chan *eventsv1.Event),
+	}
+}
+
+func (e *EventsFeed) Stop() {
+	e.cancel()
+}
+
+func (e *EventsFeed) Shutdown() {
+	e.Stop()
+	select {
+	case <-e.quit:
+	default:
+		close(e.quit)
+		close(e.feed)
```

Comment:
close, but if two threads call `Shutdown()` at the same time, they will both `close(quit)` & whoever happens to be second will blow up.

sorry :grimacing: